[CONTP-1610][CONTP-1611] Wait for CSI driver node server pod readiness in untaint controller if csi feature is enabled (#3096)
DataDog/datadog-operator: docs/untaint_controller.md

## Overview

The Untaint controller watches Kubernetes Nodes carrying the taint
`agent.datadoghq.com/not-ready=presence:NoSchedule` and removes it when
readiness criteria are met (see below), or after a configurable timeout. It is
intended to run alongside a separate mechanism (cluster-autoscaler hook, CCM,
admission webhook, etc.) that adds the taint to new nodes.

**With `--untaintControllerEnabled=true` only** (and without `--untaintControllerWaitForCSIDriver`):
the controller removes the taint once the **node Agent** pod
(`agent.datadoghq.com/component=agent`) on that node is `Ready`. Agent pods are
listed in the operator's agent watch namespaces (`WATCH_NAMESPACE` /
`DD_AGENT_WATCH_NAMESPACE`).

**With `--untaintControllerEnabled=true` and `--untaintControllerWaitForCSIDriver=true`:**
the controller waits until **both** the node Agent and **CSI
node-server** pod (`app=datadog-csi-driver-node-server`) on the node are

> `Ready` before removing the taint. The taint stays until both are

Reviewer: ```suggestion
`Ready` before removing the taint. The taint stays until both criteria are
```

> satisfied or a timeout fires. The operator's Pod informer then watches the

Reviewer: ```suggestion
met or a timeout fires. The operator's Pod informer then watches the
```

**union** of `DD_AGENT_WATCH_NAMESPACE` and `DD_CSIDRIVER_WATCH_NAMESPACE` (all

> pods in those namespaces—keep namespaces tight). Ensure CSI namespaces are

Reviewer: ```suggestion
pods in those namespaces. Keep these namespaces tightly scoped to limit the pod informer's watch scope.). CSI namespaces must be
```

covered so the controller can list CSI pod status.

**`--datadogCSIDriverEnabled`** only controls whether the **DatadogCSIDriver**
controller runs; it does **not** by itself turn on dual-readiness untaint.
Enable `--untaintControllerWaitForCSIDriver` only when you actually deploy CSI

> node-server pods on tainted nodes (for example via a `DatadogCSIDriver` CR with

Reviewer: ```suggestion
node-server pods on tainted nodes (for example, with a `DatadogCSIDriver` CR with
```

```yaml
args:
  - --untaintControllerEnabled=true
  # Optional: require CSI node-server Ready before untainting (see Overview).
  - --untaintControllerWaitForCSIDriver=true
```

| `--untaintControllerEnabled` | `--untaintControllerWaitForCSIDriver` | Behavior |
| ----------------------------- | ------------------------------------- | -------- |
| `false` | any | Untaint controller off; no Agent startup toleration for this feature. |
| `true` | `false` | Agent-only readiness; Agent DaemonSet startup toleration injected. |
| `true` | `true` | Wait for Agent **and** CSI node-server Ready; widened Pod cache (agent + `DD_CSIDRIVER_WATCH_NAMESPACE` namespaces); startup toleration on Agent and, when the DatadogCSIDriver controller is enabled, on the CSI node DaemonSet. |

> `--untaintControllerWaitForCSIDriver` requires `--untaintControllerEnabled=true` (the operator exits on invalid combinations).

Reviewer: ```suggestion
**Note**: `--untaintControllerWaitForCSIDriver` requires `--untaintControllerEnabled=true`. The operator exits at startup if this combination is invalid.
```

> When `--untaintControllerEnabled` is enabled, the operator injects a toleration for

Reviewer 2: ~~`UntaintControllerWaitForCSIDriver` this should be enabled too right, to inject tolerations?~~ misread.
